src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,322 @@
+
+        /// <summary>
+        /// Notes if we need to signal the user's unregister event after all callbacks complete.
+        /// </summary>
+        private bool _signalAfterCallbacksComplete;
+
+        private bool _unregisterCalled;
+
+        private bool _unregistered;
+
+        private AutoResetEvent _callbacksComplete;
+
+        private AutoResetEvent _removed;
+
+        /// <summary>
+        /// Unregisters this wait handle registration from the wait threads.
+        /// </summary>
+        /// <param name="waitObject">The event to signal when the handle is unregistered.</param>
+        /// <returns>If the handle was successfully marked to be removed and the provided wait handle was set as the user provided event.</returns>
+        /// <remarks>
+        /// This method will only return true on the first call.
+        /// Passing in a wait handle with a value of -1 will result in a blocking wait, where Unregister will not return until the full unregistration is completed.
+        /// </remarks>
         public bool Unregister(WaitHandle waitObject)
         {
-            // UNIXTODO: ThreadPool
-            throw new NotImplementedException();
+            GC.SuppressFinalize(this);
+            _callbackLock.Acquire();
+            bool needToRollBackRefCountOnException = false;
+            try
+            {
+                if (_unregisterCalled)
+                {
+                    return false;
+                }
+
+                UserUnregisterWaitHandle = waitObject?.SafeWaitHandle;
+                UserUnregisterWaitHandle?.DangerousAddRef();
+                needToRollBackRefCountOnException = true;
+
+                UserUnregisterWaitHandleValue = UserUnregisterWaitHandle?.DangerousGetHandle() ?? IntPtr.Zero;
+
+                if (_unregistered)
+                {
+                    SignalUserWaitHandle();
+                    return true;
+                }
+
+                if (IsBlocking)
+                {
+                    _callbacksComplete = RentEvent();
+                }
+                else
+                {
+                    _removed = RentEvent();
+                }
+                _unregisterCalled = true;
+            }
+            catch (Exception) // Rollback state on exception
+            {
+                if (_removed != null)
+                {
+                    ReturnEvent(_removed);
+                    _removed = null;
+                }
+                else if (_callbacksComplete != null)
+                {
+                    ReturnEvent(_callbacksComplete);
+                    _callbacksComplete = null;
+                }
+
+                UserUnregisterWaitHandleValue = (IntPtr)(-1);
+
+                if (needToRollBackRefCountOnException)
+                {
+                    UserUnregisterWaitHandle?.DangerousRelease();
+                }
+
+                UserUnregisterWaitHandle  = new SafeWaitHandle((IntPtr)(-1), false);
```

Comment:
I think it can just be null initially and here. The only thing we use it for is to add and remove a ref, the handle value is only used after overwriting it.